### PR TITLE
Use default SslProtocols in Kestrel

### PR DESCRIPTION
**Pull request by halter73**
Addresses #14997

I hope to get this into preview6. I will make a breaking change announcement.

### CODE REVIEW

**Comment by Tratcher** [comment]
Did you want to put it in the preview6 branch first? Or do you plan to cherry pick?

**Comment by Tratcher** [comment]
I'm a little concerned that you didn't have to update any tests...

**Comment by halter73** [comment]
> I'm a little concerned that you didn't have to update any tests...

I did. I just pushed before tests finished 😆

I'll retarget to release/5.0-preview6.

**Comment by halter73** [comment]
@blowdart Do you want us to log anything if a client connects using TLS 1.0 or SSL? If so, should we log a warning the first time this happens? On every connection where this happens?

File: src/Servers/Kestrel/test/InMemory.FunctionalTests/HttpsConnectionMiddlewareTests.cs [hunk 1]
```diff
@@ -362,12 +362,13 @@ await using (var server = new TestServer(context => context.Response.WriteAsync(
...
 
         [Fact]
-        public async Task DoesNotSupportTls10()
+        public async Task Tls10CanBeDisabled()
         {
             void ConfigureListenOptions(ListenOptions listenOptions)
             {
                 listenOptions.UseHttps(options =>
                 {
+                    options.SslProtocols = SslProtocols.Tls12 | SslProtocols.Tls12;
```
Comment (nil4) [inline src/Servers/Kestrel/test/InMemory.FunctionalTests/HttpsConnectionMiddlewareTests.cs:371]:
```suggestion
                    options.SslProtocols = SslProtocols.Tls11 | SslProtocols.Tls12;
```

Comment (blowdart) [reply src/Servers/Kestrel/test/InMemory.FunctionalTests/HttpsConnectionMiddlewareTests.cs:371]: I though we had decided to drop selection, and default to the OS like we advise everyone else to do?

**Comment by blowdart** [comment]
I don't know if we should log a warning for older protocol. How clever can we be?

I think

1) Log protocol for every connection as informational.
2) If developer has selected protocol as a setting, warn on ones which don't match.

Depending on the listener though you might not even get to 2, as the connection may be refused before then?

**Review by blowdart — CHANGES_REQUESTED** [review]
With logging, sure. I think log protocol for every connection as informational. This is useful because it'll allow a developer to gather stats so they can, for example, know when it's safe to turn an older protocol off.

If the OS is configured correctly, you wouldn't even get to the log though would you? Http.sys for example would refuse the connection.

File: src/Servers/Kestrel/test/InMemory.FunctionalTests/HttpsConnectionMiddlewareTests.cs (hunk 1 above)
Comment (nil4) [reply src/Servers/Kestrel/test/InMemory.FunctionalTests/HttpsConnectionMiddlewareTests.cs:371]: Right, but the way this is written now, this is just `SslProtocols.Tls12`. The same value was on the left *and* right side of the `|` operator, which looks like a typo, isn't it?

**Comment by halter73** [comment]
> If the OS is configured correctly, you wouldn't even get to the log though would you? Http.sys for example would refuse the connection.

Kestrel refuses the connection but still [logs the authentication failure](https://github.com/dotnet/aspnetcore/blob/15015d51626a679d851de43a5de5eb7b67200ae4/src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs#L235) at the debug level. Kestrel doesn't know why the handshake failed. It simply catches AuthenticationExceptions thrown by AuthenticateAsServerAsync and logs it.

> With logging, sure. I think log protocol for every connection as informational.

I can log the protocol for every TLS/SSL connection. I think it would be at the debug level like we do for AuthenticationExceptions though.

@blowdart If we see that TLS 1.0 or SSL is being used, should we log something different or increase the log severity?

**Comment by blowdart** [comment]
Can we make the log severity configurable? It could get noisy, and the ability to let folks choose would be useful.

**Comment by halter73** [comment]
> Can we make the log severity configurable? It could get noisy, and the ability to let folks choose would be useful.

I don't think we ever make the severity we log at configurable. We generally try to pick a reasonable level for each log and let customers configure a min log level with namespace granularity.

Debug is pretty noisy, but at least it's not trace. I'm OK with trying info if you think that's better.

File: src/Servers/Kestrel/test/InMemory.FunctionalTests/HttpsTests.cs [hunk 2]
```diff
@@ -366,7 +366,10 @@ public async Task ClientAttemptingToUseUnsupportedProtocolIsLoggedAsDebug()
                 new TestServiceContext(LoggerFactory),
                 listenOptions =>
                 {
-                    listenOptions.UseHttps(TestResources.GetTestCertificate("no_extensions.pfx"));
+                    listenOptions.UseHttps(TestResources.GetTestCertificate("no_extensions.pfx"), httpsOptions =>
+                    {
+                        httpsOptions.SslProtocols = SslProtocols.Tls12 | SslProtocols.Tls11;
```
Comment (Tratcher) [inline src/Servers/Kestrel/test/InMemory.FunctionalTests/HttpsTests.cs:371]: What happens without this?

Comment (halter73) [reply src/Servers/Kestrel/test/InMemory.FunctionalTests/HttpsTests.cs:371]: TLS 1.0 is not allowed by default on my Windows 10 Insider Preview machine. This was brought up as a problem in #14997, is why we didn't use the defaults in the first place, and is why I was asking about whether we should log at a higher level if TLS 1.1 or SSL is used.

Comment (halter73) [reply src/Servers/Kestrel/test/InMemory.FunctionalTests/HttpsTests.cs:371]: I'm deleting my last comment and resubmitting so those reading this via notifications or email will hopefully be less confused. I wrote "is not" in the last comment when I meant "is".

TLS 1.0 *is* ~not~ allowed by default on my Windows 10 Insider Preview machine. This was brought up as a problem in #14997, is why we didn't use the defaults in the first place, and is why I was asking about whether we should log at a higher level if TLS 1.0 or SSL is used.

**Comment by halter73** [comment]
@blowdart I've updated the PR to log the "SslProtocol" at the debug level after a successful handshake. This is what the logs look like at the start of a TLS 1.2 connection with debug logs enabled:

```
dbug: Microsoft.AspNetCore.Server.Kestrel[39]
      Connection id "0HM085QVC1C8G" accepted.
dbug: Microsoft.AspNetCore.Server.Kestrel[1]
      Connection id "0HM085QVC1C8G" started.
dbug: Microsoft.AspNetCore.Server.Kestrel.Https.Internal.HttpsConnectionMiddleware[3]
      Connection "0HM085QVC1C8G" established using the following protocol: Tls12
info: Microsoft.AspNetCore.Hosting.Diagnostics[1]
      Request starting HTTP/2 GET https://localhost:5001/ - -
```

Does this seem reasonable to you?

There's a lot of additional details we could log about the connection like the CipherAlgorithm, but we can address that separately. It's already possible for developers to log additional details themselves using the [ITlsHandshakeFeature](https://github.com/dotnet/aspnetcore/blob/4b713b9a53f09c0e6ba475c70b363aa19224ce56/src/Servers/Connections.Abstractions/src/Features/ITlsHandshakeFeature.cs).

**Comment by blowdart** [comment]
Works for me :) A sample for logging the rest would be good.

**Review by blowdart — APPROVED** [review]

**Comment by Pilchie** [comment]
I approve, and agree with trying to get it in early.

**Review by Tratcher — APPROVED** [review]

**Comment by halter73** [comment]
@Pilchie Can you merge this?

**Pull request merged**